### PR TITLE
flatten hierarchies feature implemented trough hold/fetch logic

### PR DESCRIPTION
**Pull request by elpie89**
so the logic here is 
hold the file
evaluate all hierarchies to check if they are collapsible or not
collapse hierarchies
fetch the file 

#557

### CODE REVIEW

**Comment by sebavan** [comment]
@elpie89 the build is failing for the max 2019 project, I leave the rest of the PR for review by @Drigax

**Comment by elpie89** [comment]
yeah sorry I messed up with the postbuild file, is ok now

**Comment by Drigax** [comment]
Looks like it works for the most part, I tried testing the flattener versus two scenes: FlattenHierarchy and FlattenHierarchySimple. I noticed that the flattener works against the simple scene, and flattens town to a single mesh per material, but it does not work against the other scene. Can you take a look at where this is failing?

File: 3ds Max/Max2Babylon/BabylonSkipFlatten.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
+﻿using System.Windows.Forms;
+using Autodesk.Max;
+using ActionItem = Autodesk.Max.Plugins.ActionItem;
+
+namespace Max2Babylon
+{
+    class BabylonInvertSkipFlatten:ActionItem
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/BabylonSkipFlatten.cs]: `class BabylonSkipFlattenFlagToggle : ActionItem`

would improve the spacing and more clear naming schema

File: 3ds Max/Max2Babylon/BabylonSkipFlatten.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        public override string MenuText
+        {
+            get
+            {
+                IINode sel = Loader.Core.GetSelNode(0);
+                if (sel == null)
+                {
+                    return "&Babylon Skip Flatten";
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/BabylonSkipFlatten.cs]: Can we change the menu text to something more intuitive like

"Scene Flattening - Enabled"
and
"Scene Flattening - Disabled"

File: 3ds Max/Max2Babylon/BabylonSkipFlatten.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                }
+
+                return "&Babylon Reset Flatten";
+            }
+        }
+
+        public override string DescriptionText
+        {
+            get { return "Invert skip flatten status"; }
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/BabylonSkipFlatten.cs]: Toggle, rather than invert

File: 3ds Max/Max2Babylon/BabylonSkipFlatten.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    class BabylonInvertSkipFlatten:ActionItem
+    {
+
+        public override bool ExecuteAction()
+        {
+            IINode sel = Loader.Core.GetSelNode(0);
+            if (sel == null) return true;
+
+            bool skip = sel.IsMarkedAsNotCollapsable();
+            sel.SetUserPropBool("babylonjs_SkipFlatten", !skip);
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/BabylonSkipFlatten.cs]: babylonjs_DoNotFlatten

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs [hunk 5]
```diff
@@ -81,7 +81,7 @@ public struct ClassIDWrapper : IEquatable<ClassIDWrapper>
         public static readonly ClassIDWrapper Switch_Shader_Material = new ClassIDWrapper(0x7e73161f, 0xa844c228);
         public static readonly ClassIDWrapper Two_Sided_Material = new ClassIDWrapper(0x7e73161f, 0x7ffd6281);
 
-
+        public static readonly ClassIDWrapper Editable_Poly = new ClassIDWrapper(469250957, 422535320);
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: I'm not sure if this is the best place to put this class IDWrapper definition, but lets leave it here for now till we move all the magic numbers somewhere logical

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs [hunk 6]
```diff
@@ -88,8 +89,103 @@ public void CheckCancelled()
                 throw new OperationCanceledException();
             }
         }
+
+        private void IsMeshCollapsable(IINode node, AnimationGroupList animationGroupList,ref List<IINode> collapsabeNodes)
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: We're using flatten in one file, and collapse in another. we should probably stick to one term for this feature.

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs [hunk 7]
```diff
@@ -88,8 +89,103 @@ public void CheckCancelled()
...
+                collapsabeINode.CollapseHierachy();
+            }
+        }
+
         public void Export(ExportParameters exportParameters)
         {
+            this.exportParameters = exportParameters;
+            IINode exportNode = (exportParameters as MaxExportParameters).exportNode;
+
+            if(exportParameters.flattenExport) FlattenHierarchy(exportNode);
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: lets not inline if statements outside of ternary operators for assignment

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs [hunk 8]
```diff
@@ -110,9 +206,8 @@ public void Export(ExportParameters exportParameters)
                 RaiseError("This parameter sets the quality of jpg compression.");
                 return;
             }
-
-            this.exportParameters = exportParameters;
-            var exportNode = (exportParameters as MaxExportParameters).exportNode;
+
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: whitespace

File: 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs [hunk 9]
```diff
@@ -1,9 +1,10 @@
...
-    }
-}
+﻿using BabylonExport.Entities;
+
+namespace Max2Babylon
+{
+    public class MaxExportParameters : ExportParameters
+    {
+        public Autodesk.Max.IINode exportNode;
+        public bool useHoldFetchLogig = true; // some script could use the babaylon exporter and override this value
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs]: is this a typo? "HoldFetchLogig"

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 10]
```diff
@@ -94,6 +94,9 @@ private void ExporterForm_Load(object sender, EventArgs e)
             string storedEnvironmentPath = Loader.Core.RootNode.GetStringProperty(ExportParameters.PBREnvironmentPathPropertyName, string.Empty);
             string formatedEnvironmentPath = Tools.ResolveRelativePath(storedEnvironmentPath);
             txtEnvironmentName.Text = formatedEnvironmentPath;
+
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: too much whitespace, remove a line

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 11]
```diff
@@ -135,7 +138,21 @@ private void btnEnvBrowse_Click(object sender, EventArgs e)
 
         private async void butExport_Click(object sender, EventArgs e)
         {
-            await DoExport(singleExportItem);
+            try
+            {
+                if (chkFlatten.Checked) Loader.Core.FileHold();
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: don't inline if statements

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 12]
```diff
@@ -187,12 +204,14 @@ private void SaveOptions()
...
             string unformattedTextureFolderPath = PathUtilities.UnformatPath(txtTextureName.Text);
-            Loader.Core.RootNode.SetStringProperty(ExportParameters.TextureFolderPathProperty,Tools.RelativePathStore(unformattedTextureFolderPath));
+            Loader.Core.RootNode.SetStringProperty(ExportParameters.TextureFolderPathProperty, Tools.RelativePathStore(unformattedTextureFolderPath));
 
             Tools.UpdateCheckBox(chkFullPBR, Loader.Core.RootNode, ExportParameters.PBRFullPropertyName);
             Tools.UpdateCheckBox(chkNoAutoLight, Loader.Core.RootNode, ExportParameters.PBRNoLightPropertyName);
             string unformattedEnvironmentPath = PathUtilities.UnformatPath(txtEnvironmentName.Text);
             Loader.Core.RootNode.SetStringProperty(ExportParameters.PBREnvironmentPathPropertyName, Tools.RelativePathStore(unformattedEnvironmentPath));
+
+            Tools.UpdateCheckBox(chkFlatten, Loader.Core.RootNode, "babylonjs_flatten");
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: can we use "babylonjs_flattenScene" as our flag name?

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 13]
```diff
@@ -295,11 +314,13 @@ private async Task<bool> DoExport(ExportItem exportItem, bool multiExport = fals
                     exportNode = exportItem != null ? exportItem.Node : null,
                     pbrNoLight = chkNoAutoLight.Checked,
                     pbrFull = chkFullPBR.Checked,
-                    pbrEnvironment = txtEnvironmentName.Text
+                    pbrEnvironment = txtEnvironmentName.Text,
+                    flattenExport = chkFlatten.Checked
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: for our exportParameter, can we use flattenScene

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 14]
```diff
@@ -399,14 +429,28 @@ private void ExporterForm_Deactivate(object sender, EventArgs e)
 
         private async void butExportAndRun_Click(object sender, EventArgs e)
         {
-            if (await DoExport(singleExportItem))
+            try
             {
-                WebServer.SceneFilename = Path.GetFileName(PathUtilities.UnformatPath(txtModelName.Text));
-                WebServer.SceneFolder = Path.GetDirectoryName(PathUtilities.UnformatPath(txtModelName.Text));
+                if (chkFlatten.Checked) Loader.Core.FileHold();
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: don't inline

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 15]
```diff
@@ -547,16 +591,29 @@ private async void butMultiExport_Click(object sender, EventArgs e)
...
                 if (chkWriteTextures.Checked || chkOverwriteTextures.Checked)
                 {
                     MessageBox.Show("Cannot write textures with Multi-File Export");
                     return;
                 }
-
-                await DoExport(exportItemList);
+                try
+                {
+                    if (chkFlatten.Checked) Loader.Core.FileHold();
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: don't inline

File: 3ds Max/Max2Babylon/GlobalUtility.cs [hunk 16]
```diff
@@ -260,10 +261,15 @@ private void InstallMenus()
             menu.AddItem(menuItemBabylon, -1);
 
             menuItemBabylon = Loader.Global.IMenuItem;
-            menuItemBabylon.Title = "Babylon Actions Builder";
+            menuItemBabylon.Title = "Babylon Invert Skip Flatten Status";
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/GlobalUtility.cs]: Can we use a better menu name here as well?

File: 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+using System.Collections.Generic;
+using System.IO;
+
+namespace Max2Babylon
+{
+    static class ScriptsUtilities
+    {
+        public static void ExecutePythonFile(string filePath)
+        {
+            //Scripts.ScriptsUtilities.ExecutePythonFile(@"hello_world.py");
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs]: remove comment

File: 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+            if (File.Exists(filePath))
+            {
+                string cmd = $@"python.ExecuteFile ""{filePath}""";
+                ExecuteMaxScriptCommand(cmd);
+            }
+        }
+
+        public static void ExecutePythonCommand(string pythonCmd)
+        {
+            //Scripts.ScriptsUtilities.ExecutePythonCommand(@"print 'helloPython'");
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs]: remove comment

File: 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+            string cmd = $@"python.Execute ""{pythonCmd}""";
+            ExecuteMaxScriptCommand(cmd);
+        }
+
+        public static void ExecuteMaxScriptCommand(string maxScriptCmd)
+        {
+            //Scripts.ScriptsUtilities.ExecuteMaxScriptCommand(@"print ""hello""");
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs]: remove comment

File: SharedProjects/Babylon2GLTF/GLTFExporter.Skin.cs [hunk 20]
```diff
@@ -42,110 +42,110 @@ private GLTFSkin ExportSkin(BabylonSkeleton babylonSkeleton, GLTF gltf, GLTFNode
...
+            // Skin
+
+            // if this mesh is sharing a skin with another mesh, use the already exported skin
+            var sharedSkinnedMeshesByOriginalPair = sharedSkinnedMeshesByOriginal.Where(skinSharingMeshPair => skinSharingMeshPair.Value.Contains(gltfMesh)).Select(kvp => (KeyValuePair<GLTFMesh, List<GLTFMesh>>?) kvp).FirstOrDefault();
+            if (sharedSkinnedMeshesByOriginalPair != null)
+            {
+                logger.RaiseMessage("GLTFExporter.Skin | Sharing skinning information from mesh '" + sharedSkinnedMeshesByOriginalPair.Value.Key.name + "'", 3);
+                var skeletonExportData = alreadyExportedSkeletons[babylonSkeleton];
+                gltfNode.skin = skeletonExportData.skinIndex;
+                return gltf.SkinsList[(int)gltfNode.skin];
```
Comment (Drigax) [inline SharedProjects/Babylon2GLTF/GLTFExporter.Skin.cs]: is this change related to the goal of the PR?

File: SharedProjects/BabylonExport.Entities/ExportParameters.cs [hunk 21]
```diff
@@ -15,6 +15,7 @@ public class ExportParameters
         public float scaleFactor = 1.0f;
         public bool writeTextures = true;
         public bool overwriteTextures = true;
+        public bool flattenExport = true;
```
Comment (Drigax) [inline SharedProjects/BabylonExport.Entities/ExportParameters.cs]: can we change this to "flattenScene"? It just gives a better idea of what it's used for. also try to homogenize usage of collapse and flatten to use flatten.

**Comment by Drigax** [comment]
Most of the issues found here are nits for code quality, but this looks pretty good!

File: SharedProjects/Babylon2GLTF/GLTFExporter.Skin.cs (hunk 20 above)
Comment (elpie89) [reply SharedProjects/Babylon2GLTF/GLTFExporter.Skin.cs]: no this was a bug, introduced after the last refactoring I think

File: 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs (hunk 19 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs]: ok

File: 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs (hunk 18 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs]: ok

File: 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs (hunk 17 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Tools/ScriptsUtilities.cs]: ok

File: 3ds Max/Max2Babylon/GlobalUtility.cs (hunk 16 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/GlobalUtility.cs]: of course, is now "Babylon Toggle Skip Flatten Status"

File: 3ds Max/Max2Babylon/BabylonSkipFlatten.cs (hunk 2 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/BabylonSkipFlatten.cs]: ok I agree

File: 3ds Max/Max2Babylon/BabylonSkipFlatten.cs (hunk 3 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/BabylonSkipFlatten.cs]: ok

File: 3ds Max/Max2Babylon/BabylonSkipFlatten.cs (hunk 4 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/BabylonSkipFlatten.cs]: of course

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs (hunk 5 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: me too, in the future, we should make a refactoring of all the constants we added in the code, including exportingParameters, userDefinedProperties etc

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs (hunk 6 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: Yes, I'm a Maya user, I'm used to name this feature collapse , in max is flatten
I replaced everything with flatten, flattenable etc.

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs (hunk 7 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: ok

File: 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs (hunk 9 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs]: I misspelled, is useHoldFetchLogic
I added this parameter because I expect that many tools could use their Hold/Fetch system
In that case, passing through MaxScript Manager the internal hold/fetch of Babylon will be overridden

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs (hunk 11 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: ok

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs (hunk 12 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: done

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs (hunk 13 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: done

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs (hunk 14 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: ok

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs (hunk 15 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: ok

File: SharedProjects/BabylonExport.Entities/ExportParameters.cs (hunk 21 above)
Comment (elpie89) [reply SharedProjects/BabylonExport.Entities/ExportParameters.cs]: done as well

File: 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs (hunk 9 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs]: I actually removed it, I  will do a second pr after to adapt the maxscript manager

**Comment by elpie89** [comment]
about the scenes that weren't working. I just changed the way the mesh is retrieved from the node and everything is ok now

File: 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs (hunk 9 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Exporter/MaxExportParameters.cs]: here some info about Hold/fetch system
https://knowledge.autodesk.com/support/3ds-max/learn-explore/caas/CloudHelp/cloudhelp/2016/ENU/3DSMax/files/GUID-4083A03C-E635-4911-9EAA-5791859EEC3E-htm.html

File: SharedProjects/Babylon2GLTF/GLTFExporter.Skin.cs (hunk 20 above)
Comment (Drigax) [reply SharedProjects/Babylon2GLTF/GLTFExporter.Skin.cs]: thanks for the fix

**Comment by Drigax** [comment]
Looks great, works great. Merge master, and I'll check it in :)

**Comment by elpie89** [comment]
done
